tensornetwork/contractors/stochastic_contractor.py

```diff
@@ -71,32 +55,39 @@ def edge_cost(edge: tensornetwork.Edge) -> int:
 
 
 def stochastic(network: tensornetwork.TensorNetwork,
-               max_rejections: int, threshold: int = -1
+               max_rejections: int, threshold: int = 1
               ) -> tensornetwork.TensorNetwork:
   """Contracts a connected network by stochastically picking edges.
+
   Algorithm 2 in page 7 of https://doi.org/10.1371/journal.pone.0208510.
   Cost calculation is slightly modified here.
+
   Args:
     network: Connected TensorNetwork to contract fully.
     max_rejections: Maximum number of rejections before you increase threshold.
     threshold: Initial value for the threshold.
+
   Returns:
     network: TensorNetwork with a single node after fully contracting.
   """
   # TODO: Think about a proper threshold value for our loss.
-  network.check_connected()
   rejections = 0
-
-  while len(network.nodes_set) > 1:
-    edge = get_random_edge(network)
-    cost = edge_cost(edge)
-    if cost <= threshold:
-      network.contract_between(edge.node1, edge.node2)
+  nondangling_edges = network.get_all_nondangling()
+  while nondangling_edges:
+    edge = random.choice(tuple(nondangling_edges))
+    if edge.node1 is edge.node2:
+      network.contract(edge)
+      nondangling_edges.remove(edge)
       rejections = 0
     else:
-      rejections += 1
-      if rejections > max_rejections:
-        threshold += 1
+      cost = edge_cost(edge)
+      if cost <= threshold:
+        network.contract_parallel(edge)
+        nondangling_edges.remove(edge)
```

Comment:
You'll have to remove all of the edges contracted from `contract_parallel`. Sadly I don't think there is an easy way to do it. :(

Reply:
Thanks for the quick review. I think an easy way is in the loop that I do in `edge_cost` (just pushed this - it was a bit quick though so can be wrong!)